MstState: fixed issues.

MBoldyrev opened the pull request:
<!-- You will not see HTML commented line in Pull Request body -->
<!-- Optional sections may be omitted. Just remove them or write None -->

<!-- ### Requirements -->
<!-- * Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion. -->
<!-- * All new code must have code coverage above 70% (https://docs.codecov.io/docs/about-code-coverage). -->
<!-- * CircleCI builds must be passed. -->
<!-- * Critical and blocker issues reported by Sorabot must be fixed. -->
<!-- * Branch must be rebased onto base branch (https://soramitsu.atlassian.net/wiki/spaces/IS/pages/11173889/Rebase+and+merge+guide). -->


### Description of the Change

The priority queue `index_` in `MstState` is expected to be sorted in the order of batches expiration, but that does not hold in general case when the queue gets sorted by the first transaction timestamp. This change sorts the queue by the oldest transaction timestamp to fix the issue.

#### UPDATE ####

- `internal_state_` and `index_` were both used to store the batches, which complicated the logic and lead to a missing cleanup operation. Now there is only one storage - `batches_`.
- `MstState::operator==(const MstState &rhs)` previously checked whether `this->internal_state_` is a subset of `rhs.internal_state_`. Implementing it like a honest equality would be ineffective with batches stored in an unordered set. As this operator is used only once in a single test, I decided better to remove it than to keep an ineffective implementation.

The index type was changed to 

<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. -->
<!-- Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->

### Benefits

Fixed `MstState::eraseByTime` and storage cleanup.

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks 

Additional boost includes present in the `mst_state.hpp` header.
Removed `MstState::operator==(const MstState &rhs)`.

<!-- What are the possible side-effects or negative impacts of the code change? -->
<!-- If no drawbacks, explicitly mention this (write None) -->

### Usage Examples or Tests *[optional]*

<!-- Point reviewers to the test, code example or documentation which shows usage example of this feature -->

### Alternate Designs *[optional]*

Rework `MstState` with pimpl to hide the includes.

<!-- Explain what other alternates were considered and why the proposed version was selected -->

MBoldyrev committed:
fixed MstState::Less

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
MstState::index_ bimap

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
MstState remove completed batches from index_

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
fixed the tests

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
fixup! MstState::index_ bimap

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
removed internal_state_

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

luckychess commented on irohad/multi_sig_transactions/state/mst_state.hpp:
```diff
@@ -164,8 +167,12 @@ namespace iroha {
                            BatchHashEquality>;
 
-    using IndexType =
-        std::priority_queue<DataType, std::vector<DataType>, Less>;
```
Looks like we don't need class `Less` anymore.

MBoldyrev committed:
Merge branch 'develop' of https://github.com/hyperledger/iroha into HEAD

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
removed MstState::operator==(const MstState &)

An effective realization would require tradeoffs. It was only used in
tests, so I optimized the rest code by the cost of uneffective equality
comparison.

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

luckychess commented on irohad/multi_sig_transactions/state/impl/mst_state.cpp:
```diff
@@ -5,15 +5,34 @@
+                      &tx) { return tx->createdTime(); });
+    const auto min_it = std::min_element(timestamps.begin(), timestamps.end());
+    assert(min_it != timestamps.end());
+    return min_it == timestamps.end() ? 0 : *min_it;
```
It looks like assertion will check only case when batch is empty so maybe better check this explicitly?
Also if empty batches are invalid here may be `optional<TimestampType>` will be a better choice? Technically 0 is a valid value (even though pretty unlikely).

MBoldyrev replied:
A batch cannot have empty transactions by external contract. I added a more explicit assertion to ensure that at the beginning of the function, as you suggested. But I want to keep this one too because I am dereferencing an iterator, which might cause a crash.

luckychess commented on test/module/irohad/multi_sig_transactions/completer_test.cpp:
```diff
@@ -99,11 +99,11 @@ TEST(CompleterTest, BatchExpirationTestExpired) {
   auto time = iroha::time::now();
   auto tx1 = std::make_shared<MockTransaction>();
-  EXPECT_CALL(*tx1, createdTime()).WillOnce(Return(time));
+  EXPECT_CALL(*tx1, createdTime()).WillRepeatedly(Return(time));
```
This and below changes look unclear, could you explain them please?

MBoldyrev replied:
Here we impose no assertions on the number of times the function must be called - it does not matter.

MBoldyrev committed:
pr fixes

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

lebdron commented on irohad/multi_sig_transactions/state/impl/mst_state.cpp:
```diff
@@ -5,15 +5,39 @@
+    auto timestamps =
+        batch->transactions()
+        | boost::adaptors::transformed(
+              +[](const std::shared_ptr<shared_model::interface::Transaction>
```
It should be possible to pass lambda to `transformed` without casting it to function pointer.

lebdron commented on irohad/multi_sig_transactions/state/impl/mst_state.cpp:
```diff
@@ -177,26 +192,23 @@ namespace iroha {
-    internal_state_.insert(rhs_batch);
-    index_.push(rhs_batch);
+    batches_.insert(
+        BatchesBimap::value_type(oldestTimestamp(rhs_batch), rhs_batch));
```
```suggestion
        {oldestTimestamp(rhs_batch), rhs_batch});
```

lebdron commented on irohad/multi_sig_transactions/state/impl/mst_state.cpp:
```diff
@@ -136,35 +150,36 @@ namespace iroha {
-        log_(std::move(log)) {}
+      : completer_(completer), log_(std::move(log)) {
+    for (const auto &batch : batches) {
+      batches_.insert(BatchesBimap::value_type(oldestTimestamp(batch), batch));
```
```suggestion
      batches_.insert({oldestTimestamp(batch), batch});
```

lebdron commented on irohad/multi_sig_transactions/state/impl/mst_state.cpp:
```diff
@@ -62,34 +82,34 @@ namespace iroha {
-        });
+    const auto &my_batches = batches_.right | boost::adaptors::map_keys;
+    std::vector<DataType> difference;
+    difference.reserve(boost::size(batches_));
```
Maybe call `shrink_to_fit` afterwards then?

MBoldyrev replied:
It would be worth shrinking if we were to keep the vector. But it is discarded with the function scope. Please see the return statement and the new constructor of `MstState`:
https://github.com/hyperledger/iroha/blob/19af4575d9047925e34a29e5effa9afcb4ab377a/irohad/multi_sig_transactions/state/mst_state.hpp#L171-L173
https://github.com/hyperledger/iroha/blob/19af4575d9047925e34a29e5effa9afcb4ab377a/irohad/multi_sig_transactions/state/mst_state.hpp#L90-L91

MBoldyrev committed:
Merge branch 'develop' into fix/mst-state-batch-comparison

MBoldyrev committed:
pr issues

Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

MBoldyrev committed:
Merge branch 'develop' into fix/mst-state-batch-comparison

MBoldyrev committed:
Merge branch 'develop' into fix/mst-state-batch-comparison
